Source/WebCore/editing/MarkupAccumulator.cpp

```diff
@@ -286,9 +305,19 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
                 break;
             }
 
-            if (m_shouldIncludeShadowDOM && current->isShadowRoot())
+            if ((m_includeShadowRoots != IncludeShadowRoots::ExplicitOnly || !m_explicitShadowRoots.isEmpty()) && current->isShadowRoot()) {
                 current = current->shadowHost();
-            else
+                if (m_includeShadowRoots != IncludeShadowRoots::DeprecatedAll) {
+                    bool shouldSkipChidren = appendContentsForNode(m_markup, *current);
+                    RefPtr currentTemplate = dynamicDowncast<HTMLTemplateElement>(*current);
+                    auto firstChild = currentTemplate ? currentTemplate->content().firstChild() : current->firstChild();
+                    if (!shouldSkipChidren && firstChild) {
+                        current = firstChild;
+                        namespaceStack.append(namespaceStack.last());
+                        break;
+                    }
```

Comment:
This is copied from code earlier in this function. Should I put it in some function that returns `Node*` which if non-null we set `current` to or is there a better way of abstracting/sharing this logic?